frontend/src/hooks/useActions.js

```diff
@@ -114,12 +123,100 @@ const useActions = (registerHotkeys = false) => {
       handler: redo
     },
     COPY: {
-      hotkey: { key: 'c', meta: true }
-      // handler: () => console.log('Copy'),
+      hotkey: { key: 'c', meta: true },
+      handler: () => {
+        const selectedStates = selectedStatesIds.map((stateId) => {
+          return project.states.find((state) => {
+            return state.id === stateId
+          })
+        })
+        const selectedComments = selectedCommentsIds.map((commentId) => {
+          return project.comments.find((comment) => {
+            return comment.id === commentId
+          })
+        })
+        const selectedTransitions = selectedTransitionsIds.map((transitionId) => {
+          return project.transitions.find((transition) => {
+            return transition.id === transitionId
+          })
+        })
+        const isInitialSelected = selectedStatesIds.includes(project.initialState)
+        // This will use the CopyData type defined in ProjectTypes
+        const copyData = {
+          states: selectedStates,
+          comments: selectedComments,
+          transitions: selectedTransitions,
+          projectSource: project._id,
+          projectType: project.projectType,
+          initialStateId: isInitialSelected ? project.initialState : null
+        }
+        navigator.clipboard.writeText(JSON.stringify(copyData))
+      }
     },
     PASTE: {
-      hotkey: { key: 'v', meta: true }
-      // handler: () => console.log('Paste'),
+      hotkey: { key: 'v', meta: true },
+      handler: async () => {
+        const pasteData = JSON.parse(await navigator.clipboard.readText())
+        let isInitialStateUpdated = false
+        if (pasteData.projectType !== project.projectType) {
+          alert(`Error: you cannot paste elements from a ${pasteData.projectType} project into a ${project.projectType} project.`)
+          return
+        }
+        const isNewProject = pasteData.projectSource !== project._id
+        // Track which transitions have been updated with new state ids
+        const newTransitions = structuredClone(pasteData.transitions)
+        newTransitions.forEach(transition => {
+          transition.from = null
+          transition.to = null
+        })
```

Comment:
I ran into a rare bug when the transitions were being updated with the new state ids of the newly created states where the ids would get updated multiple times. So I had to have a way of marking which ids had been updated and which weren't to avoid updating ids more than once. The duplicate array of transitions starts as all null and then ids are updated, so we know that if an id is not null it has already been updated to the new state id and shouldn't be touched. We still have the original transitions in pasteData so that we know what needs to be updated. If this is not clear, I am happy to explain further with an example or discuss alternative approaches because I feel like there's probably better ways to do this